File: scripts/fetch-unexotica.py
#!/usr/bin/env python3
"""Download the UnExoticA Amiga game-music collection (personal local mirror)."""

from __future__ import annotations

import json
import re
import sys
import time
import urllib.parse
from pathlib import Path

import requests

ROOT = Path(__file__).resolve().parents[1]
DEST = ROOT / "data" / "amiga" / "unexotica"
PROGRESS = DEST / ".fetch-progress.json"
WIKI_ALL = "https://www.exotica.org.uk/mediawiki/index.php?title=UnExoticA/Games_By_Title/ALL&action=raw"
WIKI_RAW = "https://www.exotica.org.uk/mediawiki/index.php"
FILE_BASE = "https://files.exotica.org.uk/"
LISTING_HOSTS = (
    "https://malus.exotica.org.uk/pub/exotica/media/audio/UnExoticA/",
    "http://malus.exotica.org.uk/pub/exotica/media/audio/UnExoticA/",
)
USER_AGENT = "RetroMusicPlayer/1.0 (personal UnExoticA mirror)"
DELAY_S = 0.2

TITLE_RE = re.compile(r"^\|\[\[([^\]]+)\]\]", re.MULTILINE)
FILE_RE = re.compile(r"\|file=(.*\.lha)\|")
BOXSCAN_RE = re.compile(r"\|boxscan=(.*\.(?:jpg|png))", re.I)
DIR_HREF = re.compile(r'href=["\']([^"\'?#]+)/["\']', re.I)
LHA_HREF = re.compile(r'href=["\']([^"\'?#]+\.lha)["\']', re.I)


def session() -> requests.Session:
    sess = requests.Session()
    sess.headers.update({"User-Agent": USER_AGENT})
    return sess


def fetch_text(sess: requests.Session, url: str, tries: int = 8) -> str | None:
    last_error: Exception | None = None
    for attempt in range(1, tries + 1):
        if attempt > 1:
            wait = min(60, 5 * attempt)
            print(f"retry {attempt}/{tries} {url} after {wait}s", flush=True)
            time.sleep(wait)
        try:
            response = sess.get(url, timeout=30)
            if response.ok and response.text:
                return response.text
            last_error = ValueError(f"HTTP {response.status_code}")
            print(f"GET {url} -> {response.status_code}", flush=True)
        except requests.RequestException as error:
            last_error = error
            print(f"GET fail {url}: {error}", flush=True)
    if last_error:
        print(f"give up {url}: {last_error}", flush=True)
    return None


def listing_dirs(html: str) -> list[str]:
    skip = {".", "..", "parent directory"}
    names: list[str] = []
    seen: set[str] = set()
    for href in DIR_HREF.findall(html):
        name = href.rstrip("/").split("/")[-1]
        if not name or name.lower() in skip or name.startswith("?"):
            continue
        if name in seen:
            continue
        seen.add(name)
        names.append(name)
    return names


def listing_lhas(html: str) -> list[str]:
    names: list[str] = []
    seen: set[str] = set()
    for href in LHA_HREF.findall(html):
        name = Path(href).name
        if name in seen:
            continue
        seen.add(name)
        names.append(name)
    return names


def crawl_game_archives(sess: requests.Session, skip_cdda: bool) -> list[tuple[str, Path]]:
    html = None
    root = None
    for host in LISTING_HOSTS:
        html = fetch_text(sess, host + "Game/")
        if html and ("href" in html.lower()):
            root = host
            break
    if not html or not root:
        return []

    composers = listing_dirs(html)
    print(f"crawl {len(composers)} composers from {root}Game/", flush=True)
    jobs: list[tuple[str, Path]] = []
    for index, composer in enumerate(composers, start=1):
        time.sleep(DELAY_S)
        page = fetch_text(sess, f"{root}Game/{composer}/", tries=4)
        if not page:
            print(f"[{index}/{len(composers)}] no listing {composer}", flush=True)
            continue
        for name in listing_lhas(page):
            if skip_cdda and "_CDDA" in name:
                continue
            rel = f"media/audio/UnExoticA/Game/{composer}/{name}"
            dest = DEST / "Game" / composer / name
            jobs.append((rel, dest))
        if index % 25 == 0 or index == len(composers):
            print(f"listed {index}/{len(composers)} composers jobs={len(jobs)}", flush=True)
    return jobs


def wiki_title(raw: str) -> str:
    if "|" in raw:
        return raw.split("|", 1)[0].strip()
    return raw.strip()


def list_titles(sess: requests.Session) -> list[str]:
    last_error: Exception | None = None
    for attempt in range(1, 61):
        if attempt > 1:
            wait = min(90, 8 * attempt)
            print(f"wiki list retry {attempt}/60 after {wait}s", flush=True)
            time.sleep(wait)
        try:
            text = sess.get(WIKI_ALL, timeout=45).text
            titles = [wiki_title(match.group(1)) for match in TITLE_RE.finditer(text)]
            if titles:
                return titles
            last_error = ValueError("empty UnExoticA title list")
        except requests.RequestException as error:
            last_error = error
            print(f"wiki list error: {error}", flush=True)
    if last_error:
        raise last_error
    return []


def archive_link(sess: requests.Session, title: str) -> str | None:
    params = {"title": title.replace(" ", "_"), "action": "raw"}
    response = sess.get(WIKI_RAW, params=params, timeout=45)
    if not response.ok:
        return None
    match = FILE_RE.search(response.text)
    return match.group(1) if match else None


def download_cover(sess: requests.Session, wiki_raw: str, archive: Path) -> bool:
    stem = archive.stem
    extract_dir = archive.parent / stem
    if (extract_dir / "cover.jpg").exists() or (extract_dir / "cover.png").exists():
        return False
    if (archive.parent / f"{stem}.jpg").exists() or (archive.parent / f"{stem}.png").exists():
        return False
    match = BOXSCAN_RE.search(wiki_raw)
    if not match:
        return False
    filename = match.group(1).strip()
    if not filename or filename.lower() == "blankboxscan.png":
        return False
    url = "https://www.exotica.org.uk/wiki/Special:Redirect/file/" + urllib.parse.quote(filename)
    response = sess.get(url, timeout=45)
    response.raise_for_status()
    data = response.content
    if data[:3] == b"\xff\xd8\xff":
        ext = ".jpg"
    elif data[:4] == b"\x89PNG":
        ext = ".png"
    else:
        return False
    extract_dir.mkdir(parents=True, exist_ok=True)
    (extract_dir / f"cover{ext}").write_bytes(data)
    (archive.parent / f"{stem}{ext}").write_bytes(data)
    return True


def is_lha(data: bytes) -> bool:
    return len(data) > 8 and b"-lh" in data[:16]


def load_progress() -> dict[str, str]:
    if not PROGRESS.exists():
        return {}
    try:
        data = json.loads(PROGRESS.read_text())
    except json.JSONDecodeError:
        return {}
    return data if isinstance(data, dict) else {}


def save_progress(progress: dict[str, str]) -> None:
    DEST.mkdir(parents=True, exist_ok=True)
    tmp = PROGRESS.with_suffix(".json.tmp")
    tmp.write_text(json.dumps(progress, indent=0, sort_keys=True))
    tmp.replace(PROGRESS)


def lha_urls(filelink: str) -> list[str]:
    suffix = filelink.split("UnExoticA/", 1)[-1]
    return [
        f"https://malus.exotica.org.uk/pub/exotica/media/audio/UnExoticA/{suffix}",
        f"http://malus.exotica.org.uk/pub/exotica/media/audio/UnExoticA/{suffix}",
        FILE_BASE + "?" + urllib.parse.urlencode({"file": "exotica/" + filelink}),
    ]


def download_lha(sess: requests.Session, filelink: str, dest: Path) -> bool:
    dest.parent.mkdir(parents=True, exist_ok=True)
    if dest.exists() and dest.stat().st_size > 32 and is_lha(dest.read_bytes()[:32]):
        return False
    last_error: Exception | None = None
    for url in lha_urls(filelink):
        try:
            response = sess.get(url, timeout=90)
            response.raise_for_status()
            data = response.content
            if is_lha(data):
                dest.write_bytes(data)
                return True
            last_error = ValueError(f"not an LHA archive from {url}")
        except requests.RequestException as error:
            last_error = error
            print(f"download fail {url}: {error}", flush=True)
            time.sleep(1)
    raise last_error or ValueError(f"not an LHA archive: {filelink}")


def extract_lha(archive: Path) -> None:
    try:
        import lhafile
    except ImportError:
        return

    out_dir = archive.parent / archive.stem
    if out_dir.exists() and any(out_dir.iterdir()):
        return

    try:
        lha = lhafile.Lhafile(str(archive))
        for info in lha.infolist():
            relative = Path(*Path(info.filename.replace("\\", "/")).parts)
            target = (out_dir / relative).resolve()
            if not str(target).startswith(str(out_dir.resolve())):
                continue
            target.parent.mkdir(parents=True, exist_ok=True)
            if relative.parts:
                target.write_bytes(lha.read(info.filename))
    except Exception as error:  # noqa: BLE001
        print(f"extract skip {archive.name}: {error}", flush=True)


def existing_lha_index() -> dict[str, Path]:
    found: dict[str, Path] = {}
    if not DEST.exists():
        return found
    for path in DEST.rglob("*.lha"):
        found[path.stem.replace("_", " ").lower()] = path
        found[path.stem.lower()] = path
    return found


def seed_progress(progress: dict[str, str], titles: list[str]) -> dict[str, str]:
    existing = existing_lha_index()
    for title in titles:
        if title in progress:
            continue
        match = existing.get(title.lower()) or existing.get(title.replace(" ", "_").lower())
        if match is None:
            continue
        progress[title] = str(match.relative_to(DEST))
        extract_lha(match)
    return progress


def crawl_main(sess: requests.Session, skip_cdda: bool) -> bool:
    jobs = crawl_game_archives(sess, skip_cdda)
    if not jobs:
        return False

    downloaded = 0
    skipped = 0
    failed = 0
    progress = load_progress()

    for index, (link, dest) in enumerate(jobs, start=1):
        key = str(dest.relative_to(DEST))
        try:
            if dest.exists() and dest.stat().st_size > 32 and is_lha(dest.read_bytes()[:32]):
                extract_lha(dest)
                progress[key] = key
                skipped += 1
                if index % 50 == 0:
                    print(f"[{index}/{len(jobs)}] exists {dest.name}", flush=True)
                continue
            time.sleep(DELAY_S)
            fresh = download_lha(sess, link, dest)
            extract_lha(dest)
            progress[key] = key
            save_progress(progress)
            if fresh:
                downloaded += 1
                print(f"[{index}/{len(jobs)}] {dest.relative_to(DEST)}", flush=True)
            else:
                skipped += 1
        except Exception as error:  # noqa: BLE001
            failed += 1
            print(f"[{index}/{len(jobs)}] FAIL {dest.name}: {error}", flush=True)

    save_progress(progress)
    print(
        f"done downloaded={downloaded} skipped={skipped} failed={failed} dest={DEST}",
        flush=True,
    )
    return True


def main() -> int:
    skip_cdda = "--cdda" not in sys.argv
    force_wiki = "--wiki" in sys.argv
    DEST.mkdir(parents=True, exist_ok=True)
    sess = session()

    if not force_wiki and crawl_main(sess, skip_cdda):
        return 0

    print("falling back to wiki title list", flush=True)
    titles = list_titles(sess)
    print(f"UnExoticA games: {len(titles)}", flush=True)

    downloaded = 0
    skipped = 0
    failed = 0
    progress = seed_progress(load_progress(), titles)

    for index, title in enumerate(titles, start=1):
        try:
            cached = progress.get(title)
            if cached == "CDDA":
                skipped += 1
                continue
            if cached:
                dest = DEST / cached
                if dest.exists() and dest.stat().st_size > 32 and is_lha(dest.read_bytes()[:32]):
                    extract_lha(dest)
                    skipped += 1
                    try:
                        raw = sess.get(
                            WIKI_RAW,
                            params={"title": title.replace(" ", "_"), "action": "raw"},
                            timeout=45,
                        ).text
                        download_cover(sess, raw, dest)
                    except Exception:
                        pass
                    if index % 50 == 0:
                        print(f"[{index}/{len(titles)}] exists {dest.name}", flush=True)
                    continue

            time.sleep(DELAY_S)
            raw_response = sess.get(
                WIKI_RAW,
                params={"title": title.replace(" ", "_"), "action": "raw"},
                timeout=45,
            )
            if not raw_response.ok:
                print(f"[{index}/{len(titles)}] no file for {title}", flush=True)
                failed += 1
                continue
            match = FILE_RE.search(raw_response.text)
            link = match.group(1) if match else None
            if not link:
                print(f"[{index}/{len(titles)}] no file for {title}", flush=True)
                failed += 1
                continue
            if skip_cdda and "_CDDA" in link:
                progress[title] = "CDDA"
                save_progress(progress)
                skipped += 1
                continue
            dest = DEST / Path(link.removeprefix("media/audio/UnExoticA/"))
            time.sleep(DELAY_S)
            fresh = download_lha(sess, link, dest)
            extract_lha(dest)
            download_cover(sess, raw_response.text, dest)
            progress[title] = str(dest.relative_to(DEST))
            save_progress(progress)
            if fresh:
                downloaded += 1
                print(f"[{index}/{len(titles)}] {dest.relative_to(DEST)}", flush=True)
            else:
                skipped += 1
                if index % 50 == 0:
                    print(f"[{index}/{len(titles)}] exists {dest.name}", flush=True)
        except Exception as error:  # noqa: BLE001
            failed += 1
            print(f"[{index}/{len(titles)}] FAIL {title}: {error}", flush=True)

    save_progress(progress)

    print(
        f"done downloaded={downloaded} skipped={skipped} failed={failed} dest={DEST}",
        flush=True,
    )
    return 0 if failed == 0 else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/fetch-unexotica.mjs
#!/usr/bin/env node
/**
 * Download UnExoticA game packs and Exotica box scans one title at a time.
 * Exotica currently serves a JS "Verifying..." gate that plain HTTP clients fail.
 */

import { spawnSync } from "node:child_process";
import {
  existsSync,
  mkdirSync,
  readFileSync,
  readdirSync,
  statSync,
  writeFileSync,
} from "node:fs";
import { dirname, join, relative } from "node:path";
import { fileURLToPath } from "node:url";
import { chromium } from "playwright";

const ROOT = join(dirname(fileURLToPath(import.meta.url)), "..");
const DEST = join(ROOT, "data", "amiga", "unexotica");
const PROGRESS = join(DEST, ".fetch-progress.json");
const PYTHON = join(ROOT, "scripts", ".venv", "bin", "python");
const LETTERS = ["0-9", ..."ABCDEFGHIJKLMNOPQRSTUVWXYZ".split("")];
const SKIP_CDDA = !process.argv.includes("--cdda");
const FILTER = process.argv.find((arg) => arg.startsWith("--filter="))?.slice(9) ?? "";
const DELAY_MS = 250;

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function isLha(buf) {
  return buf.length > 8 && buf.subarray(0, 16).includes(Buffer.from("-lh"));
}

function imageExt(buf) {
  if (buf.length > 3 && buf[0] === 0xff && buf[1] === 0xd8 && buf[2] === 0xff) return ".jpg";
  if (
    buf.length > 8 &&
    buf[0] === 0x89 &&
    buf[1] === 0x50 &&
    buf[2] === 0x4e &&
    buf[3] === 0x47
  ) {
    return ".png";
  }
  return null;
}

function coverPath(dir) {
  const jpg = join(dir, "cover.jpg");
  const png = join(dir, "cover.png");
  if (existsSync(jpg) && statSync(jpg).size > 32) return jpg;
  if (existsSync(png) && statSync(png).size > 32) return png;
  return null;
}

function coverForArchive(archive) {
  const dir = dirname(archive);
  const stem = archive.replace(/\.lha$/i, "").split("/").pop() ?? "cover";
  const extractDir = join(dir, stem);
  return (
    coverPath(extractDir) ||
    (existsSync(join(dir, `${stem}.jpg`)) ? join(dir, `${stem}.jpg`) : null) ||
    (existsSync(join(dir, `${stem}.png`)) ? join(dir, `${stem}.png`) : null)
  );
}

function loadProgress() {
  if (!existsSync(PROGRESS)) return {};
  try {
    const data = JSON.parse(readFileSync(PROGRESS, "utf8"));
    return data && typeof data === "object" ? data : {};
  } catch {
    return {};
  }
}

function saveProgress(progress) {
  mkdirSync(DEST, { recursive: true });
  writeFileSync(PROGRESS, JSON.stringify(progress, null, 0));
}

function existingLhaIndex() {
  const found = new Map();
  if (!existsSync(DEST)) return found;
  const stack = [DEST];
  while (stack.length) {
    const dir = stack.pop();
    for (const name of readdirSync(dir)) {
      const path = join(dir, name);
      const stat = statSync(path);
      if (stat.isDirectory()) {
        stack.push(path);
        continue;
      }
      if (!name.toLowerCase().endsWith(".lha")) continue;
      const stem = name.slice(0, -4);
      found.set(stem.replaceAll("_", " ").toLowerCase(), path);
      found.set(stem.toLowerCase(), path);
    }
  }
  return found;
}

function extractLha(archive) {
  if (!existsSync(PYTHON)) return;
  const script = `
import sys
from pathlib import Path
archive = Path(sys.argv[1])
try:
    import lhafile
except ImportError:
    sys.exit(0)
out_dir = archive.parent / archive.stem
if out_dir.exists() and any(out_dir.iterdir()):
    sys.exit(0)
try:
    lha = lhafile.Lhafile(str(archive))
    for info in lha.infolist():
        relative = Path(*Path(info.filename.replace(chr(92), "/")).parts)
        target = (out_dir / relative).resolve()
        if not str(target).startswith(str(out_dir.resolve())):
            continue
        target.parent.mkdir(parents=True, exist_ok=True)
        if relative.parts:
            target.write_bytes(lha.read(info.filename))
except Exception as error:
    print(f"extract skip {archive.name}: {error}", flush=True)
`;
  spawnSync(PYTHON, ["-c", script, archive], { stdio: "inherit" });
}

function isVerifyPage(text) {
  return /Verifying your browser/i.test(text) || /<title>Verifying/i.test(text);
}

async function waitVerified(page, url = "https://www.exotica.org.uk/wiki/UnExoticA") {
  await page.goto(url, { waitUntil: "domcontentloaded", timeout: 45000 });
  await page.waitForFunction(() => !/verifying/i.test(document.title), { timeout: 15000 });
}

async function requestBuffer(context, page, url) {
  let response = await context.request.get(url, { timeout: 90000 });
  let buf = Buffer.from(await response.body());
  const head = buf.toString("utf8", 0, 400);
  if (!isVerifyPage(head)) return buf;
  await waitVerified(page);
  response = await context.request.get(url, { timeout: 90000 });
  return Buffer.from(await response.body());
}

async function openWiki(page, url) {
  await page.goto(url, { waitUntil: "domcontentloaded", timeout: 45000 });
  if (/verifying/i.test(await page.title())) {
    await page.waitForFunction(() => !/verifying/i.test(document.title), { timeout: 15000 });
  }
}

async function listTitles(page) {
  const titles = [];
  const seen = new Set();
  for (const letter of LETTERS) {
    await sleep(DELAY_MS);
    await openWiki(page, `https://www.exotica.org.uk/wiki/UnExoticA/Games_By_Title/${letter}`);
    const found = await page.evaluate(() => {
      const content = document.querySelector("#mw-content-text");
      if (!content) return [];
      return [...content.querySelectorAll("table tr")]
        .map((row) => {
          const cell = row.querySelector("td:first-child a");
          return (cell?.getAttribute("title") || cell?.textContent || "").trim();
        })
        .filter((name) => name && !name.includes("Games By Title"));
    });
    let count = 0;
    for (const title of found) {
      if (seen.has(title)) continue;
      seen.add(title);
      titles.push(title);
      count += 1;
    }
    console.log(`letter ${letter}: ${count} titles`);
  }
  return titles;
}

function archiveUrl(filelink) {
  const params = new URLSearchParams({ file: `exotica/${filelink}` });
  return `http://files.exotica.org.uk/?${params.toString()}`;
}

async function downloadLha(context, filesPage, filelink, dest) {
  mkdirSync(dirname(dest), { recursive: true });
  if (existsSync(dest) && statSync(dest).size > 32 && isLha(readFileSync(dest).subarray(0, 32))) {
    return false;
  }
  const url = archiveUrl(filelink);
  try {
    const buf = await requestBuffer(context, filesPage, url);
    if (isLha(buf)) {
      writeFileSync(dest, buf);
      return true;
    }
  } catch {
    // Fall through to a real navigation, which can pass the JS gate.
  }
  const downloadPromise = filesPage.waitForEvent("download", { timeout: 60000 });
  // Always settle so a failed goto cannot leave an unhandled rejection.
  const ignoreDownload = downloadPromise.then(
    () => undefined,
    () => undefined,
  );
  try {
    await filesPage.goto(url, { waitUntil: "domcontentloaded", timeout: 45000 });
    if (/verifying/i.test(await filesPage.title())) {
      await filesPage.waitForFunction(() => !/verifying/i.test(document.title), { timeout: 15000 });
      await filesPage.reload({ waitUntil: "domcontentloaded" });
    }
    const download = await downloadPromise;
    await download.saveAs(dest);
  } catch (error) {
    await ignoreDownload;
    throw error;
  }
  const saved = readFileSync(dest).subarray(0, 32);
  if (!isLha(saved)) {
    throw new Error(`not an LHA archive (${statSync(dest).size} bytes)`);
  }
  return true;
}

function isDirectImageUrl(url) {
  return /\/mediawiki\/files\/.+\.(jpe?g|png)(\?|$)/i.test(url);
}

async function downloadCover(context, page, assets, archivePath) {
  if (coverForArchive(archivePath)) return false;
  let url = assets.coverHref || assets.coverSrc || null;
  if (!url || /blankboxscan/i.test(url)) return false;
  if (!isDirectImageUrl(url)) {
    await sleep(DELAY_MS);
    await openWiki(page, url);
    url = await page.evaluate(() => {
      const img = document.querySelector("#file img, .fullImageLink img, img.mw-file-element");
      return img ? img.src : null;
    });
  }
  if (!url) url = assets.coverSrc || null;
  if (!url || (!isDirectImageUrl(url) && !/\.(jpe?g|png)(\?|$)/i.test(url))) return false;
  await sleep(DELAY_MS);
  const buf = await requestBuffer(context, page, url);
  const ext = imageExt(buf);
  if (!ext) return false;
  const dir = dirname(archivePath);
  const stem = archivePath.replace(/\.lha$/i, "").split("/").pop() ?? "cover";
  const extractDir = join(dir, stem);
  mkdirSync(extractDir, { recursive: true });
  writeFileSync(join(extractDir, `cover${ext}`), buf);
  writeFileSync(join(dir, `${stem}${ext}`), buf);
  return true;
}

function readWikiAssets() {
  const lha = [...document.querySelectorAll("a")].find(
    (a) => /\.lha/i.test(a.getAttribute("href") || "") || /\.lha/i.test(a.textContent || ""),
  );
  let link = null;
  if (lha) {
    try {
      const file = new URL(lha.href).searchParams.get("file");
      if (file) link = file.replace(/^exotica\//, "");
      else if (/Game\/.+\.lha$/i.test((lha.textContent || "").trim())) {
        link = `media/audio/UnExoticA/${lha.textContent.trim()}`;
      }
    } catch {
      link = null;
    }
  }
  const box = [...document.querySelectorAll("a")].find(
    (a) =>
      /box scan/i.test(a.getAttribute("title") || "") ||
      /box scan/i.test(a.textContent || ""),
  );
  const coverImg = [...document.querySelectorAll("img")].find((img) =>
    /box scan/i.test(img.alt || ""),
  );
  const disambiguation = [...document.querySelectorAll("#mw-content-text a")].find((a) =>
    /\(game\)\s*$/i.test((a.textContent || "").trim()),
  );
  return {
    link,
    coverHref: box && !/blank/i.test(box.href) ? box.href : null,
    coverSrc: coverImg && !/blank/i.test(coverImg.src) ? coverImg.src : null,
    gameHref: disambiguation ? disambiguation.href : null,
  };
}

async function wikiAssets(page, title) {
  const slug = title.replaceAll(" ", "_");
  await openWiki(page, `https://www.exotica.org.uk/wiki/${encodeURIComponent(slug)}`);
  let assets = await page.evaluate(readWikiAssets);
  if (!assets.link && !assets.coverHref && !assets.coverSrc && assets.gameHref) {
    await openWiki(page, assets.gameHref);
    assets = await page.evaluate(readWikiAssets);
  }
  return assets;
}

function knownArchive(progress, existing, title) {
  const cached = progress[title];
  if (cached && cached !== "CDDA") return join(DEST, cached);
  const bare = title.replace(/\s*\([^)]+\)\s*$/, "").trim();
  const keys = [title, bare].flatMap((value) => [
    value.toLowerCase(),
    value.replaceAll(" ", "_").toLowerCase(),
  ]);
  for (const key of keys) {
    const hit = existing.get(key);
    if (hit) return hit;
  }
  return undefined;
}

async function main() {
  mkdirSync(DEST, { recursive: true });
  const existing = existingLhaIndex();
  const progress = loadProgress();
  const browser = await chromium.launch({ headless: true });
  const context = await browser.newContext({
    userAgent:
      "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36",
  });
  const page = await context.newPage();
  const filesPage = await context.newPage();

  console.log("passing Exotica browser check…");
  await waitVerified(page, "https://www.exotica.org.uk/wiki/UnExoticA");

  let titles;
  const exact = FILTER.match(/^\^(.+)\$$/);
  if (exact) {
    titles = [exact[1].replaceAll("\\", "")];
    console.log(`UnExoticA games: 1 (${titles[0]})`);
  } else {
    titles = await listTitles(page);
    if (FILTER) {
      const filter = new RegExp(FILTER, "i");
      titles = titles.filter((title) => filter.test(title));
    }
    console.log(`UnExoticA games: ${titles.length}`);
  }

  let downloaded = 0;
  let skipped = 0;
  let failed = 0;
  let covers = 0;

  for (const [index, title] of titles.entries()) {
    const n = index + 1;
    try {
      const cached = progress[title];
      if (cached === "CDDA") {
        skipped += 1;
        continue;
      }
      const known = knownArchive(progress, existing, title);
      const haveLha =
        known &&
        existsSync(known) &&
        statSync(known).size > 32 &&
        isLha(readFileSync(known).subarray(0, 32));

      if (haveLha) {
        extractLha(known);
        progress[title] = relative(DEST, known);
        skipped += 1;
        if (!coverForArchive(known)) {
          await sleep(DELAY_MS);
          const assets = await wikiAssets(page, title);
          if (await downloadCover(context, page, assets, known)) {
            covers += 1;
            console.log(`[${n}/${titles.length}] cover ${relative(DEST, dirname(known))}`);
          }
        }
        if (n % 50 === 0) console.log(`[${n}/${titles.length}] exists ${known.split("/").pop()}`);
        continue;
      }

      await sleep(DELAY_MS);
      const assets = await wikiAssets(page, title);
      const link = assets.link;
      if (!link) {
        console.log(`[${n}/${titles.length}] no file for ${title}`);
        failed += 1;
        continue;
      }
      if (SKIP_CDDA && link.includes("_CDDA")) {
        progress[title] = "CDDA";
        saveProgress(progress);
        skipped += 1;
        continue;
      }
      const dest = join(DEST, link.replace(/^media\/audio\/UnExoticA\//, ""));
      await sleep(DELAY_MS);
      const fresh = await downloadLha(context, filesPage, link, dest);
      extractLha(dest);
      if (await downloadCover(context, page, assets, dest)) covers += 1;
      progress[title] = relative(DEST, dest);
      saveProgress(progress);
      if (fresh) {
        downloaded += 1;
        console.log(`[${n}/${titles.length}] ${relative(DEST, dest)}`);
      } else {
        skipped += 1;
      }
    } catch (error) {
      failed += 1;
      console.log(`[${n}/${titles.length}] FAIL ${title}: ${error.message}`);
    }
  }

  saveProgress(progress);
  await browser.close();
  console.log(
    `done downloaded=${downloaded} covers=${covers} skipped=${skipped} failed=${failed} dest=${DEST}`,
  );
  process.exit(failed === 0 ? 0 : 1);
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});
